fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -2136,92 +2131,92 @@ void populateBitset(boost::dynamic_bitset<>& bs, std::vector<uint16_t>& ids) {
+	std::unordered_map<Version, boost::dynamic_bitset<>> versionAvailableTLogs;
+	// version -> all tLogs that the version was sent to (by the commit proxy)
+	std::map<Version, boost::dynamic_bitset<>> versionAllTLogs;
+	// version -> prevVersion (that was given out by the sequencer) map
+	std::map<Version, Version> prevVersionMap;
+	uint16_t tLogIdx = 0;
+	int replicationFactor = std::get<0>(logGroupResults);
+	for (auto& tLogResult : std::get<1>(logGroupResults)) {
+		if (tLogResult.unknownCommittedVersions.empty()) {
+			return minKCV;
+		}
+		uint16_t tLogLocId = tLogLocIds[tLogIdx];
+		availableTLogs.set(tLogLocId);
+		for (auto& unknownCommittedVersion : tLogResult.unknownCommittedVersions) {
+			Version k = unknownCommittedVersion.version;
+			if (k > minKCV) {
+				if (versionAvailableTLogs[k].empty()) {
+					versionAvailableTLogs[k].resize(bsSize);
 				}
+				versionAvailableTLogs[k].set(tLogLocId);
+				prevVersionMap[k] = unknownCommittedVersion.prev;
+				if (versionAllTLogs[k].empty()) {
+					versionAllTLogs[k].resize(bsSize);
+				}
+				populateBitset(versionAllTLogs[k], unknownCommittedVersion.tLogLocIds);
 			}
-			if (!logGroupCandidate) {
-				return minEnd;
-			}
-			tLogIdx++;
 		}
-		Version minTLogs = minEnd;
-		Version prevVersion = minEnd;
-		for (auto const& [version, tLogs] : versionAllTLogs) {
-			if (!(prevVersion == minEnd || prevVersion == prevVersionMap[version])) {
-				break;
-			}
-			// This version is not recoverable if there is a log server (LS) such that:
-			// - the commit proxy sent this version to LS (i.e., LS is present in "versionAllTLogs[version]")
-			// - LS is available (i.e., LS is present in "availableTLogs")
-			// - LS didn't receive this version (i.e., LS is not present in "versionAvailableTLogs[version]")
-			if (((tLogs & availableTLogs) & ~versionAvailableTLogs[version]).any()) {
-				break;
-			}
-			// If the commit proxy sent this version to "N" log servers then at least
-			// (N - replicationFactor + 1) log servers must be available.
-			if (!(versionAvailableTLogs[version].size() >= tLogs.size() - replicationFactor + 1)) {
-				break;
-			}
-			// Update RV.
-			minTLogs = version;
-			// Update recovery version vector.
-			for (boost::dynamic_bitset<>::size_type id = 0; id < versionAvailableTLogs[version].size(); id++) {
-				if (versionAvailableTLogs[version][id]) {
-					RVs[id] = version;
-				}
+		tLogIdx++;
+	}
+
+	// Compute recovery version.
+	// @note we think that the unicast recovery version should always greater than or
+	// equal to "min(DV)" (= "minDV"). To be conservative we use "min(KCV)" (= "minKCV")
+	// as the default (starting) recovery version and later verify that the computed
+	// recovery version is greater than or equal to "minDV".
+	// @todo modify code to use "minDV" as the default (starting) recovery version
+	// @todo to be investigated: maybe we should use max(KCV) as the starting recovery
```

Comment:
I think this will result in a "DBRecoveryDurabilityError" error in simulation tests. Let's see if the tests hit this case (assuming this is a real issue).